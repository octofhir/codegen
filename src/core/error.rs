//! Error types for codegen

use thiserror::Error;

/// Result type alias
pub type Result<T> = std::result::Result<T, Error>;

/// Main error type for codegen operations
#[derive(Error, Debug)]
pub enum Error {
    /// IO error
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    /// JSON parsing error
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),

    /// TOML parsing error
    #[error("TOML error: {0}")]
    Toml(#[from] toml::de::Error),

    /// Canonical manager error
    #[error("Canonical manager error: {0}")]
    CanonicalManager(String),

    /// Parser error
    #[error("Parser error: {0}")]
    Parser(String),

    /// Generator error
    #[error("Generator error: {0}")]
    Generator(String),

    /// Validation error
    #[error("Validation error: {0}")]
    Validation(String),

    /// Configuration error
    #[error("Configuration error: {0}")]
    Config(String),

    /// Template error
    #[error("Template error: {0}")]
    Template(String),

    /// Other errors
    #[error("{0}")]
    Other(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_types() {
        let err = Error::Parser("test error".to_string());
        assert_eq!(err.to_string(), "Parser error: test error");
    }

    #[test]
    fn test_error_from_io() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let err: Error = io_err.into();
        assert!(err.to_string().contains("IO error"));
    }
}
