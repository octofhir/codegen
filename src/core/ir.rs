//! Intermediate Representation (IR) for FHIR types
//!
//! This module defines language-agnostic representations of FHIR resources,
//! datatypes, and primitives that can be transformed into any target language.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Complete type graph representing all FHIR definitions
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TypeGraph {
    /// Resource definitions (e.g., Patient, Observation)
    pub resources: IndexMap<String, ResourceType>,

    /// Complex datatype definitions (e.g., HumanName, Address)
    pub datatypes: IndexMap<String, DataType>,

    /// Primitive type definitions (e.g., string, boolean, date)
    pub primitives: IndexMap<String, PrimitiveType>,

    /// Profile definitions (e.g., USCorePatient)
    pub profiles: IndexMap<String, ProfileType>,

    /// FHIR version this graph represents
    pub fhir_version: FhirVersion,

    /// Metadata about generation
    pub metadata: GraphMetadata,
}

impl TypeGraph {
    /// Create a new empty type graph
    pub fn new(fhir_version: FhirVersion) -> Self {
        Self {
            resources: IndexMap::new(),
            datatypes: IndexMap::new(),
            primitives: IndexMap::new(),
            profiles: IndexMap::new(),
            fhir_version,
            metadata: GraphMetadata::default(),
        }
    }

    /// Add a resource to the graph
    pub fn add_resource(&mut self, name: String, resource: ResourceType) {
        self.resources.insert(name, resource);
    }

    /// Add a datatype to the graph
    pub fn add_datatype(&mut self, name: String, datatype: DataType) {
        self.datatypes.insert(name, datatype);
    }

    /// Add a primitive to the graph
    pub fn add_primitive(&mut self, name: String, primitive: PrimitiveType) {
        self.primitives.insert(name, primitive);
    }

    /// Add a profile to the graph
    pub fn add_profile(&mut self, name: String, profile: ProfileType) {
        self.profiles.insert(name, profile);
    }

    /// Get total number of types in graph
    pub fn total_types(&self) -> usize {
        self.resources.len() + self.datatypes.len() + self.primitives.len() + self.profiles.len()
    }
}

/// Metadata about the type graph
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GraphMetadata {
    /// When this graph was generated
    pub generated_at: String,

    /// Generator version
    pub generator_version: String,

    /// Source packages
    pub source_packages: Vec<String>,

    /// Custom metadata
    pub custom: HashMap<String, serde_json::Value>,
}

impl Default for GraphMetadata {
    fn default() -> Self {
        Self {
            generated_at: chrono::Utc::now().to_rfc3339(),
            generator_version: env!("CARGO_PKG_VERSION").to_string(),
            source_packages: Vec::new(),
            custom: HashMap::new(),
        }
    }
}

/// FHIR version
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum FhirVersion {
    /// FHIR R4 (4.0.x)
    R4,
    /// FHIR R4B (4.3.x)
    R4B,
    /// FHIR R5 (5.0.x)
    R5,
    /// FHIR R6 (6.0.x)
    R6,
}

impl std::fmt::Display for FhirVersion {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FhirVersion::R4 => write!(f, "R4"),
            FhirVersion::R4B => write!(f, "R4B"),
            FhirVersion::R5 => write!(f, "R5"),
            FhirVersion::R6 => write!(f, "R6"),
        }
    }
}

/// FHIR Resource type (e.g., Patient, Observation)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResourceType {
    /// Resource name (e.g., "Patient")
    pub name: String,

    /// Base resource this extends (e.g., Some("DomainResource"))
    pub base: Option<String>,

    /// Properties/elements of this resource
    pub properties: Vec<Property>,

    /// Search parameters defined for this resource
    pub search_parameters: Vec<SearchParameter>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Complex datatype (e.g., HumanName, Address)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DataType {
    /// Datatype name (e.g., "HumanName")
    pub name: String,

    /// Base type this extends (if any)
    pub base: Option<String>,

    /// Properties of this datatype
    pub properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Primitive type (e.g., string, boolean, date)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PrimitiveType {
    /// Primitive name (e.g., "string", "boolean")
    pub name: String,

    /// Base primitive (all primitives extend from Element)
    pub base: Option<String>,

    /// Regex pattern for validation (if any)
    pub pattern: Option<String>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// FHIR Profile (constraint on a resource or datatype)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProfileType {
    /// Profile name (e.g., "USCorePatient")
    pub name: String,

    /// Base type being constrained (e.g., "Patient")
    pub base: String,

    /// Additional constraints on properties
    pub property_constraints: Vec<PropertyConstraint>,

    /// New properties added by this profile
    pub new_properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// Property constraint in a profile
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PropertyConstraint {
    /// Path to the property being constrained
    pub path: String,

    /// New cardinality (if changed)
    pub cardinality: Option<CardinalityRange>,

    /// New type constraints
    pub type_constraints: Vec<String>,

    /// New binding (if changed)
    pub binding: Option<ValueSetBinding>,

    /// Must support flag
    pub must_support: bool,
}

/// A property (field/element) in a resource or datatype
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Property {
    /// Property name (e.g., "name", "birthDate")
    pub name: String,

    /// Original FHIR path (e.g., "Patient.name")
    pub path: String,

    /// Type of this property
    pub property_type: PropertyType,

    /// Cardinality (min..max)
    pub cardinality: CardinalityRange,

    /// Is this a choice element? (e.g., value[x])
    pub is_choice: bool,

    /// If choice, what are the possible types?
    pub choice_types: Vec<String>,

    /// Is this a modifier element?
    pub is_modifier: bool,

    /// Is this a summary element?
    pub is_summary: bool,

    /// Terminology binding (if any)
    pub binding: Option<ValueSetBinding>,

    /// Constraints/invariants
    pub constraints: Vec<InvariantRule>,

    /// Short description
    pub short_description: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Example values
    pub examples: Vec<Example>,
}

/// Type of a property
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum PropertyType {
    /// Primitive type (e.g., "string", "boolean")
    Primitive {
        /// Type name
        type_name: String,
    },

    /// Complex type (e.g., "HumanName", "Address")
    Complex {
        /// Type name
        type_name: String,
    },

    /// Reference to another resource
    Reference {
        /// Allowed target types (empty = any resource)
        target_types: Vec<String>,
    },

    /// Backbone element (inline complex type)
    BackboneElement {
        /// Properties in this backbone element
        properties: Vec<Property>,
    },

    /// Choice type (one of several types)
    Choice {
        /// Possible types for this choice
        types: Vec<String>,
    },
}

impl PropertyType {
    /// Get the primary type name (for simple cases)
    pub fn type_name(&self) -> Option<&str> {
        match self {
            PropertyType::Primitive { type_name } => Some(type_name),
            PropertyType::Complex { type_name } => Some(type_name),
            _ => None,
        }
    }
}

/// Cardinality range (min..max)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct CardinalityRange {
    /// Minimum occurrences
    pub min: u32,

    /// Maximum occurrences (None = unbounded)
    pub max: Option<u32>,
}

impl CardinalityRange {
    /// Create a required single value (1..1)
    pub fn required() -> Self {
        Self { min: 1, max: Some(1) }
    }

    /// Create an optional single value (0..1)
    pub fn optional() -> Self {
        Self { min: 0, max: Some(1) }
    }

    /// Create a required array (1..*)
    pub fn required_array() -> Self {
        Self { min: 1, max: None }
    }

    /// Create an optional array (0..*)
    pub fn optional_array() -> Self {
        Self { min: 0, max: None }
    }

    /// Is this property required?
    pub fn is_required(&self) -> bool {
        self.min >= 1
    }

    /// Is this property an array?
    pub fn is_array(&self) -> bool {
        self.max.is_none() || self.max.is_some_and(|max| max > 1)
    }

    /// Is this property optional?
    pub fn is_optional(&self) -> bool {
        self.min == 0
    }
}

/// ValueSet binding for coded elements
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValueSetBinding {
    /// Binding strength
    pub strength: BindingStrength,

    /// ValueSet canonical URL
    pub value_set: String,

    /// Description
    pub description: Option<String>,
}

/// Binding strength
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum BindingStrength {
    /// Required (must use)
    Required,
    /// Extensible (should use)
    Extensible,
    /// Preferred (could use)
    Preferred,
    /// Example (may use)
    Example,
}

/// Invariant constraint rule
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InvariantRule {
    /// Invariant key (e.g., "pat-1")
    pub key: String,

    /// Severity
    pub severity: ConstraintSeverity,

    /// Human description
    pub human: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// XPath expression (legacy)
    pub xpath: Option<String>,
}

/// Constraint severity
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ConstraintSeverity {
    /// Error (must satisfy)
    Error,
    /// Warning (should satisfy)
    Warning,
}

/// Example value
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Example {
    /// Example label
    pub label: String,

    /// Example value (as JSON)
    pub value: serde_json::Value,
}

/// Documentation for types and properties
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct Documentation {
    /// Short description
    pub short: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Requirements
    pub requirements: Option<String>,

    /// Usage notes
    pub usage_notes: Vec<String>,

    /// Canonical URL
    pub url: Option<String>,
}

/// Search parameter definition
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SearchParameter {
    /// Parameter name/code
    pub code: String,

    /// Parameter type
    pub param_type: SearchParamType,

    /// Description
    pub description: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// Target resource types (for reference params)
    pub target_types: Vec<String>,
}

/// Search parameter type
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SearchParamType {
    /// Number
    Number,
    /// Date/DateTime
    Date,
    /// String
    String,
    /// Token
    Token,
    /// Reference
    Reference,
    /// Composite
    Composite,
    /// Quantity
    Quantity,
    /// URI
    Uri,
    /// Special
    Special,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_type_graph_creation() {
        let graph = TypeGraph::new(FhirVersion::R4);
        assert_eq!(graph.fhir_version, FhirVersion::R4);
        assert_eq!(graph.total_types(), 0);
    }

    #[test]
    fn test_type_graph_add_resource() {
        let mut graph = TypeGraph::new(FhirVersion::R4);

        let resource = ResourceType {
            name: "Patient".to_string(),
            base: Some("DomainResource".to_string()),
            properties: Vec::new(),
            search_parameters: Vec::new(),
            documentation: Documentation::default(),
            url: "http://hl7.org/fhir/StructureDefinition/Patient".to_string(),
            is_abstract: false,
        };

        graph.add_resource("Patient".to_string(), resource);
        assert_eq!(graph.total_types(), 1);
        assert!(graph.resources.contains_key("Patient"));
    }

    #[test]
    fn test_cardinality_required() {
        let card = CardinalityRange::required();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, Some(1));
        assert!(card.is_required());
        assert!(!card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_cardinality_optional() {
        let card = CardinalityRange::optional();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, Some(1));
        assert!(!card.is_required());
        assert!(!card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_array() {
        let card = CardinalityRange::optional_array();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, None);
        assert!(!card.is_required());
        assert!(card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_required_array() {
        let card = CardinalityRange::required_array();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, None);
        assert!(card.is_required());
        assert!(card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_property_type_typename() {
        let prim = PropertyType::Primitive { type_name: "string".to_string() };
        assert_eq!(prim.type_name(), Some("string"));

        let complex = PropertyType::Complex { type_name: "HumanName".to_string() };
        assert_eq!(complex.type_name(), Some("HumanName"));

        let reference = PropertyType::Reference { target_types: vec!["Patient".to_string()] };
        assert_eq!(reference.type_name(), None);
    }

    #[test]
    fn test_fhir_version_display() {
        assert_eq!(FhirVersion::R4.to_string(), "R4");
        assert_eq!(FhirVersion::R4B.to_string(), "R4B");
        assert_eq!(FhirVersion::R5.to_string(), "R5");
        assert_eq!(FhirVersion::R6.to_string(), "R6");
    }

    #[test]
    fn test_graph_metadata_default() {
        let metadata = GraphMetadata::default();
        assert_eq!(metadata.generator_version, env!("CARGO_PKG_VERSION"));
        assert!(metadata.source_packages.is_empty());
        assert!(!metadata.generated_at.is_empty());
    }
}
