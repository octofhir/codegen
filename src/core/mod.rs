//! Core functionality for codegen

pub mod error;
pub mod graph_builder;
pub mod ir;
pub mod parser;
pub mod resolver;

pub use error::{Error, Result};
pub use graph_builder::TypeGraphBuilder;
pub use ir::TypeGraph;
pub use parser::StructureDefinitionParser;
pub use resolver::SchemaResolver;
