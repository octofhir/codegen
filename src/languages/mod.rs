//! Language-specific code generation backends
//!
//! This module contains implementations for different target languages.

pub mod typescript;
