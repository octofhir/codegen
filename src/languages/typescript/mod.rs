//! TypeScript code generation

mod backend;
pub mod choice_generator;
pub mod datatype_generator;
/// Documentation generation for TypeScript with JSDoc and TypeDoc support
pub mod documentation_generator;
/// Helper methods generation for TypeScript
pub mod helpers_generator;
/// Package manifest generation for TypeScript
pub mod manifest_generator;
pub mod resource_generator;
/// Complete TypeScript SDK generation orchestrator
pub mod sdk_generator;
pub mod templates;
/// Validation functions generation for TypeScript
pub mod validation_generator;

pub use backend::TypeScriptBackend;
pub use choice_generator::ChoiceElementGenerator;
pub use datatype_generator::DatatypeGenerator;
pub use documentation_generator::DocumentationGenerator;
pub use helpers_generator::HelpersGenerator;
pub use manifest_generator::{ManifestGenerator, PackageConfig};
pub use resource_generator::ResourceGenerator;
pub use sdk_generator::TypeScriptSdkGenerator;
pub use validation_generator::ValidationGenerator;
