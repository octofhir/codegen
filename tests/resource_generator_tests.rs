use insta::assert_snapshot;
use octofhir_codegen::core::ir::*;
use octofhir_codegen::languages::typescript::{ResourceGenerator, TypeScriptBackend};

fn create_patient_resource() -> ResourceType {
    ResourceType {
        name: "Patient".to_string(),
        base: Some("DomainResource".to_string()),
        properties: vec![
            Property {
                name: "identifier".to_string(),
                path: "Patient.identifier".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "Identifier".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "An identifier for this patient".to_string(),
                definition: "An identifier for this patient".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "active".to_string(),
                path: "Patient.active".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "boolean".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: true,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Whether this patient's record is in active use".to_string(),
                definition: "Whether this patient's record is in active use".to_string(),
                comments: Some("If a record is inactive, and linked to an active record, then future patient/record updates should occur on the other patient.".to_string()),
                examples: vec![],
            },
            Property {
                name: "name".to_string(),
                path: "Patient.name".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "HumanName".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "A name associated with the patient".to_string(),
                definition: "A name associated with the individual".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "telecom".to_string(),
                path: "Patient.telecom".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "ContactPoint".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "A contact detail for the individual".to_string(),
                definition: "A contact detail (e.g. a telephone number or an email address) by which the individual may be contacted.".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "gender".to_string(),
                path: "Patient.gender".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/administrative-gender".to_string(),
                    description: Some("The gender of a person used for administrative purposes.".to_string()),
                }),
                constraints: vec![],
                short_description: "male | female | other | unknown".to_string(),
                definition: "Administrative Gender - the gender that the patient is considered to have for administration and record keeping purposes.".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "birthDate".to_string(),
                path: "Patient.birthDate".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "date".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "The date of birth for the individual".to_string(),
                definition: "The date of birth for the individual".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "deceased".to_string(),
                path: "Patient.deceased[x]".to_string(),
                property_type: PropertyType::Choice {
                    types: vec!["boolean".to_string(), "dateTime".to_string()],
                },
                cardinality: CardinalityRange::optional(),
                is_choice: true,
                choice_types: vec!["boolean".to_string(), "dateTime".to_string()],
                is_modifier: true,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Indicates if the individual is deceased or not".to_string(),
                definition: "Indicates if the individual is deceased or not".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "address".to_string(),
                path: "Patient.address".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "Address".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "An address for the individual".to_string(),
                definition: "An address for the individual".to_string(),
                comments: None,
                examples: vec![],
            },
        ],
        search_parameters: vec![],
        documentation: Documentation {
            short: "Information about an individual or animal receiving care or other health-related services".to_string(),
            definition: "Demographics and other administrative information about an individual or animal receiving care or other health-related services.".to_string(),
            url: Some("http://hl7.org/fhir/StructureDefinition/Patient".to_string()),
            ..Default::default()
        },
        url: "http://hl7.org/fhir/StructureDefinition/Patient".to_string(),
        is_abstract: false,
    }
}

fn create_observation_resource() -> ResourceType {
    ResourceType {
        name: "Observation".to_string(),
        base: Some("DomainResource".to_string()),
        properties: vec![
            Property {
                name: "identifier".to_string(),
                path: "Observation.identifier".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "Identifier".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Business Identifier for observation".to_string(),
                definition: "A unique identifier assigned to this observation".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "status".to_string(),
                path: "Observation.status".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::required(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: true,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/observation-status".to_string(),
                    description: Some("Codes providing the status of an observation.".to_string()),
                }),
                constraints: vec![],
                short_description: "registered | preliminary | final | amended +".to_string(),
                definition: "The status of the result value".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "code".to_string(),
                path: "Observation.code".to_string(),
                property_type: PropertyType::Complex {
                    type_name: "CodeableConcept".to_string(),
                },
                cardinality: CardinalityRange::required(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Type of observation (code / type)".to_string(),
                definition: "Describes what was observed. Sometimes this is called the observation \"name\".".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "subject".to_string(),
                path: "Observation.subject".to_string(),
                property_type: PropertyType::Reference {
                    target_types: vec!["Patient".to_string(), "Group".to_string(), "Device".to_string(), "Location".to_string()],
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Who and/or what the observation is about".to_string(),
                definition: "The patient, or group of patients, location, or device this observation is about".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "value".to_string(),
                path: "Observation.value[x]".to_string(),
                property_type: PropertyType::Choice {
                    types: vec![
                        "Quantity".to_string(),
                        "CodeableConcept".to_string(),
                        "string".to_string(),
                        "boolean".to_string(),
                        "integer".to_string(),
                        "Range".to_string(),
                    ],
                },
                cardinality: CardinalityRange::optional(),
                is_choice: true,
                choice_types: vec![
                    "Quantity".to_string(),
                    "CodeableConcept".to_string(),
                    "string".to_string(),
                    "boolean".to_string(),
                    "integer".to_string(),
                    "Range".to_string(),
                ],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Actual result".to_string(),
                definition: "The information determined as a result of making the observation".to_string(),
                comments: None,
                examples: vec![],
            },
        ],
        search_parameters: vec![],
        documentation: Documentation {
            short: "Measurements and simple assertions".to_string(),
            definition: "Measurements and simple assertions made about a patient, device or other subject.".to_string(),
            url: Some("http://hl7.org/fhir/StructureDefinition/Observation".to_string()),
            ..Default::default()
        },
        url: "http://hl7.org/fhir/StructureDefinition/Observation".to_string(),
        is_abstract: false,
    }
}

#[test]
fn test_generate_patient_resource() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);
    let resource = create_patient_resource();

    let output = generator.generate_resource(&resource).unwrap();
    assert_snapshot!("patient_resource", output);
}

#[test]
fn test_generate_observation_resource() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);
    let resource = create_observation_resource();

    let output = generator.generate_resource(&resource).unwrap();
    assert_snapshot!("observation_resource", output);
}

#[test]
fn test_generate_patient_type_guard() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);

    let output = generator.generate_type_guard("Patient").unwrap();
    assert_snapshot!("patient_type_guard", output);
}

#[test]
fn test_generate_observation_type_guard() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);

    let output = generator.generate_type_guard("Observation").unwrap();
    assert_snapshot!("observation_type_guard", output);
}

#[test]
fn test_generate_resource_union() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);

    let mut graph = TypeGraph::new(FhirVersion::R4);
    graph.add_resource("Patient".to_string(), create_patient_resource());
    graph.add_resource("Observation".to_string(), create_observation_resource());

    let output = generator.generate_resource_union(&graph).unwrap();
    assert_snapshot!("resource_union", output);
}

#[test]
fn test_generate_all_type_guards() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);

    let mut graph = TypeGraph::new(FhirVersion::R4);
    graph.add_resource("Patient".to_string(), create_patient_resource());
    graph.add_resource("Observation".to_string(), create_observation_resource());

    let output = generator.generate_all_type_guards(&graph).unwrap();
    assert_snapshot!("all_type_guards", output);
}

#[test]
fn test_generate_resource_constants() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);

    let mut graph = TypeGraph::new(FhirVersion::R4);
    graph.add_resource("Patient".to_string(), create_patient_resource());
    graph.add_resource("Observation".to_string(), create_observation_resource());

    let output = generator.generate_resource_constants(&graph).unwrap();
    assert_snapshot!("resource_constants", output);
}

#[test]
fn test_collect_patient_dependencies() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);
    let resource = create_patient_resource();

    let deps = generator.collect_dependencies(&resource);

    assert!(deps.contains("Identifier"));
    assert!(deps.contains("HumanName"));
    assert!(deps.contains("ContactPoint"));
    assert!(deps.contains("Address"));
    assert!(deps.contains("DomainResource"));
    assert!(!deps.contains("boolean")); // Primitives should not be included
    assert!(!deps.contains("string"));
    assert!(!deps.contains("date"));
}

#[test]
fn test_collect_observation_dependencies() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);
    let resource = create_observation_resource();

    let deps = generator.collect_dependencies(&resource);

    assert!(deps.contains("Identifier"));
    assert!(deps.contains("CodeableConcept"));
    assert!(deps.contains("DomainResource"));
    // Reference type should include target types - but currently it doesn't have type_name()
    // Choice types won't be in deps since type_name() returns None for Choice
}

#[test]
fn test_generate_patient_file_with_imports() {
    let backend = TypeScriptBackend::new();
    let generator = ResourceGenerator::new(backend);
    let resource = create_patient_resource();

    let deps = generator.collect_dependencies(&resource);
    let output = generator.generate_resource_file(&resource, &deps).unwrap();

    assert_snapshot!("patient_file_complete", output);
}
