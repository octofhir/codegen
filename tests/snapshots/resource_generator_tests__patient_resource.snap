---
source: tests/resource_generator_tests.rs
expression: output
---
/**
 * Information about an individual or animal receiving care or other health-related services
 * 
 * Demographics and other administrative information about an individual or animal
 * receiving care or other health-related services.
 * 
 * @see http://hl7.org/fhir/StructureDefinition/Patient
 */
export interface Patient extends DomainResource {
  resourceType: "Patient";
  identifier?: Identifier[];
  active?: boolean;
  name?: HumanName[];
  telecom?: ContactPoint[];
  gender?: string;
  birthDate?: string;
  deceased?: boolean | dateTime;
  address?: Address[];
}
